### PR TITLE
Start to add ssh key attest

### PR DESCRIPTION
**Pull request by Firstyear**
This is the start of ssh key attestation, but due to https://bugzilla.mindrot.org/show_bug.cgi?id=3516 it is not currently possible to proceed.

### CODE REVIEW

**Comment by Firstyear** [comment]
@micolous This is the start of ssh key attestation, it has all the needed bits now and I've worked out how the ssh command line works to actually do the challenges properly. I think this ties back to some of the MDS work so should likely wait for that before I commit it. 

Saying this, I'm not 100% sure if this belongs in this library or a seperate crate that consumers webauthn-rs-core. What do you think? The other option is to feature gate it.

File: webauthn-rs-core/Cargo.toml [hunk 1]
```diff
@@ -18,13 +18,14 @@ default = []
...
 [dependencies]
+base64.workspace = true
 base64urlsafedata.workspace = true
 webauthn-rs-proto.workspace = true
 serde.workspace = true
 serde_cbor.workspace = true
 serde_json.workspace = true
 nom.workspace = true
-base64.workspace = true
+sshkeys = { path = "/Users/william/development/rust-sshkeys", features = ["serde"] }
```
Comment (micolous) [inline webauthn-rs-core/Cargo.toml:28]: Please fix

Comment (Firstyear) [reply webauthn-rs-core/Cargo.toml:28]: https://github.com/dnaeon/rust-sshkeys/issues/9 waiting on 0.3.3 :)

**Comment by Firstyear** [comment]
@micolous Broke out to it's own lib, added the dep as a git ref.

File: sshkey-attest/src/proto.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+use webauthn_rs_core::attestation::AttestationFormat;
+use webauthn_rs_core::proto::{ParsedAttestation, RegisteredExtensions};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+/*
+#[serde(
+    try_from = "SerialisableAttestedPublicKey",
+    into = "SerialisableAttestedPublicKey"
+)]
+*/
```
Comment (micolous) [inline sshkey-attest/src/proto.rs:14]: Is there any reason to serialise this struct? Can this commented attribute can be deleted?

File: webauthn-rs-core/src/crypto.rs [hunk 3]
```diff
@@ -143,9 +144,13 @@ where
     }
 }
 
+/// TpmSanData
```
Comment (micolous) [inline webauthn-rs-core/src/crypto.rs:147]: I feel like this is to silence a warning for undocumented public features, but this is already public, and this comment doesn't really explain what this structure is.

File: webauthn-rs-core/src/crypto.rs [hunk 4]
```diff
@@ -302,9 +307,9 @@ pub(crate) fn assert_tpm_attest_req(x509: &x509::X509) -> Result<(), WebauthnErr
     Ok(())
 }
 
-pub(crate) fn assert_packed_attest_req(pubk: &x509::X509) -> Result<(), WebauthnError> {
-    // Verify that attestnCert meets the requirements in § 8.2.1 Packed Attestation
-    // Statement Certificate Requirements.
+/// Verify that attestnCert meets the requirements in § 8.2.1 Packed Attestation
```
Comment (micolous) [inline webauthn-rs-core/src/crypto.rs:310]: Link to https://www.w3.org/TR/webauthn-2/#sctn-packed-attestation-cert-requirements here

File: webauthn-rs-core/src/error.rs [hunk 5]
```diff
@@ -274,6 +274,21 @@ pub enum WebauthnError {
...
     #[error("The attestation ca list can not be empty")]
     MissingAttestationCaList,
+
+    #[error("This key has an invalid backup state flag")]
+    SshPublicKeyBackupState,
+
+    #[error("This key is not a valid ssh public key")]
+    SshPublicKeyInvalidType,
+
+    #[error("This requested ssh public key curve is invalid")]
```
Comment (micolous) [inline webauthn-rs-core/src/error.rs:284]: "The requested"

File: webauthn-rs-core/src/error.rs [hunk 6]
```diff
@@ -274,6 +274,21 @@ pub enum WebauthnError {
...
+    SshPublicKeyInvalidCurve,
+
+    #[error("This public key string is not a valid public key")]
```
Comment (micolous) [inline webauthn-rs-core/src/error.rs:287]: "The SSH public key is invalid"

(though I'm not conflicted as to whether to include the word "string" there)

File: webauthn-rs-core/src/internals.rs [hunk 7]
```diff
@@ -1092,16 +1095,16 @@ fn parse_tpmupublicid(input: &[u8], alg: TpmAlgId) -> nom::IResult<&[u8], TpmuPu
 
 #[derive(Debug)]
 /// Tpm Public Key Structure
-pub struct TpmtPublic {
+pub(crate) struct TpmtPublic {
     /// The type of public parms and key IE Ecdsa or Rsa
-    pub type_: TpmAlgId,
+    pub _type_: TpmAlgId,
```
Comment (micolous) [inline webauthn-rs-core/src/internals.rs:1100]: `_type` should be sufficient

File: sshkey-attest/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+        i,
+        SshSkAttestationRaw {
+            att_cert_raw,
+            sig_raw,
+            auth_data_raw,
+        },
+    ))
+}
+
+fn to_ssh_pubkey(cose: COSEKey) -> Result<PublicKey, WebauthnError> {
```
Comment (micolous) [inline sshkey-attest/src/lib.rs:303]: Could this be `impl TryFrom<CoseKey> for PublicKey`?

File: sshkey-attest/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+            att_cert,
+            sig,
+            // Probably need auth_data raw.
+            auth_data_bytes,
+            auth_data,
+        })
+    }
+}
+
+fn ssh_sk_attestation_parser(i: &[u8]) -> nom::IResult<&[u8], SshSkAttestationRaw> {
```
Comment (micolous) [inline sshkey-attest/src/lib.rs:275]: Name as a verb: `parse_ssh_sk_attestation`

File: sshkey-attest/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+        }
+        None => ExtnState::NotRequested,
+    };
+
+    let extensions = RegisteredExtensions {
+        cred_protect,
+        ..Default::default()
+    };
+
+    // Assert that backup eligible and state are both false.
```
Comment (micolous) [inline sshkey-attest/src/lib.rs:182]: > Assert that the key is not backup-eligible and not backed up.

File: sshkey-attest/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+    attestation_cas: &AttestationCaList,
+    danger_disable_certificate_time_checks: bool,
+) -> Result<AttestedPublicKey, WebauthnError> {
+    if attestation_cas.is_empty() {
+        return Err(WebauthnError::MissingAttestationCaList);
+    }
+
+    let alg = COSEAlgorithm::ES256;
+
+    // There doesn't seem to be much in the way of docs about the format of
```
Comment (micolous) [inline sshkey-attest/src/lib.rs:79]: This comment seems to be about `ssh_sk_attestation_parser`; and should be in that function.

Also, it looks like there is documentation: https://github.com/openssh/openssh-portable/blob/2709809fd616a0991dc18e3a58dea10fb383c3f0/PROTOCOL.u2f#L151-L174

File: sshkey-attest/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+        }
+        _ => {
+            debug!("to_ssh_pubkey");
+            Err(WebauthnError::SshPublicKeyInvalidType)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (micolous) [inline sshkey-attest/src/lib.rs:352]: Can there be some tests to check for expected attestation failures, seeing as that's security-sensitive?

File: sshkey-attest/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+#![deny(clippy::needless_pass_by_value)]
+#![deny(clippy::trivially_copy_pass_by_ref)]
+
+#[macro_use]
+extern crate tracing;
+
+pub mod proto;
+
+use uuid::Uuid;
+use webauthn_rs_core::attestation::verify_attestation_ca_chain;
```
Comment (micolous) [inline sshkey-attest/src/lib.rs:25]: Merge `use` statements

File: sshkey-attest/src/proto.rs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+/// An attested public key. This contains the ssh public key as well as the
+/// attestation metadata.
+pub struct AttestedPublicKey {
+    /// The ssh public key
+    pub pubkey: PublicKey,
+    /// The set of registrations that were verified at registration, that can
```
Comment (micolous) [inline sshkey-attest/src/proto.rs:20]: "set of extensions which"

File: sshkey-attest/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+
+use crate::proto::AttestedPublicKey;
+
+/// Given an attestation generated by ssh-keygen, parse and validate it's content
+/// returning an attested public key structure that contains metadata and the ssh
+/// public key.
+///
+/// You can create an attested sk key with:
+/// ```shell
+///  dd if=/dev/urandom of=/Users/username/.ssh/id_ecdsa_sk.chal bs=16 count=1
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply sshkey-attest/src/lib.rs:60]: ssh-keygen doesn't do shell expansion.

File: sshkey-attest/src/lib.rs (hunk 8 above)
Comment (Firstyear) [reply sshkey-attest/src/lib.rs:303]: Yes, but that then makes it public since you can't control visibility of impl trait.

**Review by micolous — APPROVED** [review]

**Pull request merged**